src/java/org/apache/cassandra/repair/SystemDistributedKeyspace.java

```diff
@@ -307,6 +320,55 @@ public static void setViewRemoved(String keyspaceName, String viewName)
         forceBlockingFlush(VIEW_BUILD_STATUS);
     }
 
+    /**
+     * Reads blacklisted partitions from system_distributed.blacklisted_partitions table.
+     * Stops reading partitions upon exceeding the cache size limit by logging a warning.
+     * @return
+     */
+    public static Set<BlacklistedPartition> getBlacklistedPartitions()
+    {
+        String query = "SELECT keyspace_name, columnfamily_name, partition_key FROM %s.%s";
+        UntypedResultSet results;
+        try
+        {
+            results = QueryProcessor.execute(format(query, SchemaConstants.DISTRIBUTED_KEYSPACE_NAME, BLACKLISTED_PARTITIONS),
+                                             ConsistencyLevel.ONE);
+        }
+        catch (Exception e)
+        {
+            logger.error("Error querying blacklisted partitions", e);
+            return Collections.emptySet();
+        }
+
+        Set<BlacklistedPartition> blacklistedPartitions = new HashSet<>();
+        int cacheSize = 0;
+        for (UntypedResultSet.Row row : results)
+        {
+            try
+            {
+                BlacklistedPartition blacklistedPartition = new BlacklistedPartition(row.getString("keyspace_name"), row.getString("columnfamily_name"), row.getString("partition_key"));
+
+                // check if adding this blacklisted partition would increase cache size beyond the set limit.
+                cacheSize += blacklistedPartition.unsharedHeapSize();
+                if (cacheSize / (1024 * 1024) > DatabaseDescriptor.getBlackListedPartitionsCacheSizeLimitInMB())
```

Comment:
It makes more sense to put this logic in `BlacklistedPartitionCache` since it is specific to caching, this also makes your caching logic easily testable.

Reply:
makes sense!